common/devdoc/sm_requirements.md

```diff
@@ -197,13 +209,18 @@ static SM_RESULT sm_close_begin_internal(SM_HANDLE sm, ON_SM_CLOSING_COMPLETE_CA
 
 **SRS_SM_02_051: [** If the state is `SM_OPENED_DRAINING_TO_BARRIER` then `sm_close_begin_internal` shall re-evaluate the state. **]**
 
+If the state is `SM_OPENING` and the `close_while_opening_callback` is non-NULL then ...
+
+... `sm_close_begin_internal` shall call the `close_while_opening_callback` and then re-evaluate the state.
```

Comment:
>m_close_begin_internal` shall call the `close_while_opening_callback` and then re-evaluate the state. [](http://example.com/codeflow?start=6&length=101)

I just realized that this means if the callback doesn't actually return the state of sm to something other than OPENING, it will keep getting called (possibly forever).

I wonder if we should either only call the callback once and just fail if still in opening the next time, or we should have a return value on the callback so the callback handler can fail if it's not possible to get out of the OPENING state.

I'm leaning toward the former because it seems like the callback should either be able to resolve this or not the first time, but maybe there's a case for "retrying" that callback (I just don't see it).

I just feel that it might help to have some way to avoid getting stuck.